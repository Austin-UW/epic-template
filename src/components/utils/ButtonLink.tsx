import React from 'react'
import { Button } from '@material-ui/core'
import { Link } from 'react-router-dom'

type ButtonProps = {
  variant: 'contained' | 'outlined'
  to: string
  text: string
  style?: object
}
/** atom */
export const ButtonLink = (props: ButtonProps) => (
  <Button
    component={(linkProps: any) => <Link to={props.to} {...linkProps} />}
    variant={props.variant}
    color="primary"
    size="large"
    style={typeof props.style === 'object' ? props.style : {}}
  >
    {props.text}
  </Button>
)
