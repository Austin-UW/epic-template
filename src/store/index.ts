export { authUserU } from './thunks/auth'

export { startLoadingA, stopLoadingA } from './actions/loading'
export { openSnackbarA, closeSnackbarA } from './actions/snackbar'
export { createProjectA } from './actions/project'

export { reducer } from './reducers/reducer'
export { store } from './store'
export { defaultState } from './defaultState'
