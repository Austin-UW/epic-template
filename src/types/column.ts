export type TColumn = {
  readonly id: number
  name: string
  isCompletedColumn?: boolean
  taskIds: number[]
}
export type TColumns = { [id: string]: TColumn }
