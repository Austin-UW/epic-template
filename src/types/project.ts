export type TBaseProject = {
  // used in create-project form and doesn't contain super fancy stuff
  name: string
  categories: string[]
}

export type TProject = TBaseProject & {
  readonly id: number
  columnIds: number[]
}

export type TProjects = { [id: string]: TProject }
