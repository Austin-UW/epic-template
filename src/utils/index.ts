import { cloneDeep } from 'lodash'

export const formalize = (str: string): string => {
  return str
    .toLowerCase()
    .replace(/,/g, '')
    .replace(/ /g, '')
}

export const filterItemsFromIds = <T = any>(
  ids: number[],
  items: { [id: string]: any } & T // add index signature
): T => {
  const newItems = cloneDeep(items) // fuckin changed in past :()
  Object.keys(newItems)
    .map(key => parseInt(key, 10)) // Object.keys returns strings nomatter what
    .forEach((key: any) => {
      if (!ids.includes(key)) {
        delete newItems[key]
      }
    })
  return newItems
}

export const getHighestId = (obj: { [i: number]: any }): number => {
  const result = Math.max(...Object.keys(obj).map(key => parseInt(key, 10)))
  if (result === -Infinity) {
    return -1
  }
  return result
}

export const getHighestIdElement = (obj: { [i: number]: any }) => {
  const highestId = getHighestId(obj)
  const highestIdElement = obj[highestId] // return the value of the highest

  return highestIdElement
}

export const getDaysInMonth = (month: number, year: number): Date[] => {
  // Since no month has fewer than 28 days
  const date: Date = new Date(year, month, 1)
  const days: Date[] = []
  while (date.getMonth() === month) {
    days.push(new Date(date))
    date.setDate(date.getDate() + 1)
  }
  return days
}
