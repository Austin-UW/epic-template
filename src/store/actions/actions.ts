import { TAction, TUser } from 'src/types'

export const setCurrentUserA = (user: TUser): TAction => {
  return {
    type: 'SET_CURRENT_USER',
    user
  }
}
export const toggleA = (open: boolean): TAction => ({ type: 'TOGGLE_MENU', open })
