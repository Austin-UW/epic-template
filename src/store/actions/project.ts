import { TBaseProject, TProject } from '../../types/project'
import { TCreateProject, TDeleteProject, TAction } from './types'
import { TState } from 'src/types/state'
import {
  filterItemsFromIds,
  getHighestIdElement
  // getHighestId
} from '../../utils'
import { TColumns } from 'src/types/column'
import { deleteColumnA, deleteColumn, createColumnA } from './column'
import { Dispatch } from 'redux'
import { createTaskA, editTaskA } from './task'
import { TTask } from 'src/types/task'

export const createCopiedProjectA = (
  project: TBaseProject,
  cloningId: number
) => {
  return (dispatch: Dispatch<TAction>, getState: () => TState) => {
    const cloning = getState().projects[cloningId]
    const cloningsColumns = Object.values(getState().columns).filter(col => {
      return cloning.columnIds.includes(col.id)
    })
    const cloningTasks: Array<TTask & { colId: number }> = Object.values(
      getState().tasks
    )
      .map(task => {
        for (const cloningCol of cloningsColumns) {
          if (cloningCol.taskIds.includes(task.id)) {
            return { ...task, colId: cloningCol.id, invalid: false } as any
          }
        }
        return { ...task }
      })
      .filter(task => task.colId !== undefined)

    dispatch(createProjectA(project))
    const newProject: TProject = getHighestIdElement(getState().projects)

    newProject.columnIds.map(columnId => {
      console.log('deleting newprojects col', columnId)
      dispatch(deleteColumn(columnId))
    })

    cloningsColumns.map(column => {
      console.log('creating cloned column: ', column.id)
      dispatch(createColumnA(column.name, newProject.id))
    })

    console.log(cloningTasks)

    cloningTasks.map(cloningTask => {
      const oldColId = cloningTask.colId
      const oldColIndex = cloning.columnIds.indexOf(oldColId)
      const firstColOfNew = newProject.columnIds[0]
      dispatch(createTaskA('', firstColOfNew + oldColIndex))
      const newTask = getHighestIdElement(getState().tasks)
      dispatch(editTaskA(newTask.id, { ...cloningTask, id: newTask.id }))
    })
  }
}

export const createProjectA = (project: TBaseProject): TCreateProject => ({
  type: 'CREATE_PROJECT',
  project
})

const deleteProject = (id: number): TDeleteProject => ({
  type: 'DELETE_PROJECT',
  id
})

export const deleteProjectA = (id: number) => {
  return (dispatch: Dispatch<TAction>, getState: () => TState) => {
    const projectsColumns: TColumns = filterItemsFromIds<TColumns>(
      getState().projects[id].columnIds,
      getState().columns
    )
    Object.keys(projectsColumns).map(key => {
      dispatch(deleteColumnA(parseInt(key, 10)) as any) // will this work?
    })
    // DO AFTER
    dispatch(deleteProject(id))
  }
}

export const renameProjectA = (name: string, id: number): TAction => ({
  type: 'RENAME_PROJECT',
  name,
  id
})
