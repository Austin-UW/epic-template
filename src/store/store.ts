import { createStore, applyMiddleware, Store } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'
import thunk from 'redux-thunk'
import { reducer } from '.'
import { defaultState } from './defaultState'
import { TState } from '../types/state'
import { incrementDailyTimeA } from './actions/pomodoro'
import { TAction } from '@store/actions/types'

export const loadState = (): undefined | TState => {
  try {
    const serializedState = localStorage.getItem('newerState')
    if (serializedState === null) {
      return undefined
    }
    const state: TState = JSON.parse(serializedState)
    Object.keys(state.tasks).map(id => {
      const task = state.tasks[id]
      if (task.dueDate) {
        const date = new Date()
        date.setTime(Date.parse(state.tasks[id].dueDate as any))
        state.tasks[id].dueDate = date
      }
    })
    return state
  } catch (err) {
    return undefined
  }
}

const saveState = (state: TState) => {
  state = {
    ...state,
    pomodoro: { ...state.pomodoro, paused: true }
  }
  try {
    const serializedState = JSON.stringify(state)
    localStorage.setItem('newerState', serializedState)
  } catch (err) {}
}

export const store: Store<TState, TAction> = createStore(
  reducer,
  defaultState,
  composeWithDevTools(applyMiddleware(thunk))
)

store.subscribe(() => {
  const pastAction = store.getState().pastAction
  const pastState = store.getState()
  if (pastAction.type === 'TICK') {
    if (pastState.pomodoro.currSeconds % 60 === 0) {
      store.dispatch(incrementDailyTimeA())
      store.dispatch({ type: 'CHECK_TASK_OVERDUE' })
      saveState(pastState) // save pomodoro time left every minute
    }
  } else if (pastAction.type === 'TICK_STOPWATCH') {
    if (pastState.pomodoro.stopWatch.time % 60 === 0) {
      store.dispatch({ type: 'CHECK_TASK_OVERDUE' })
      saveState(pastState)
    }
  } else if (pastAction.type !== 'CHECK_TASK_OVERDUE') {
    // dont let itself call itself that not good
    // store.dispatch({ type: 'CHECK_TASK_OVERDUE' })
    saveState(pastState)
  }
})
