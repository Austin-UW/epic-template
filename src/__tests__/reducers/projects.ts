import { defaultState, createProjectA, store } from '../../store'
import { baseProjectStandard } from '../data/baseProject'
import { getHighestId } from '../../utils'
import { renameProjectA, deleteProjectA } from '../../store/actions/project'
import { getHighestIdElement } from '../../utils'
import { createColumnA, moveColumnA } from '../../store/actions/column'

// we will create, rename, and delete a project
describe('project reducer', () => {
  test('defaults to correct state', () => {
    expect(store.getState().projects).toEqual(defaultState.projects)
  })
  test('CREATE_PROJECT', () => {
    const highestElemColIds: number[] = getHighestIdElement(
      store.getState().projects
    ).columnIds
    const id: number = highestElemColIds[highestElemColIds.length - 1]

    store.dispatch(createProjectA(baseProjectStandard))
    expect(getHighestIdElement(store.getState().projects)).toEqual({
      ...baseProjectStandard,
      columnIds: [id, id + 1, id + 2],
      id: getHighestId(store.getState().projects)
    })
  })
  test('RENAME_PROJECT', () => {
    store.dispatch(
      renameProjectA(
        'new name',
        getHighestIdElement(store.getState().projects).id
      )
    )
    expect(getHighestIdElement(store.getState().projects).name).toBe('new name')
  })
  test('DELETE_PROJECT', () => {
    const oldId = 0
    store.dispatch(deleteProjectA(oldId) as any)
    expect(store.getState().projects[0]).toBeUndefined()
    expect(store.getState().columns['2']).toBeUndefined()
    expect(store.getState().columns['3']).toBeDefined()
    expect(store.getState().tasks['0']).toBeUndefined()
  })
  test('CREATE_COLUMN', () => {
    const prev = store.getState().projects[2].columnIds.length
    store.dispatch(createColumnA('NAME', 2, true))
    expect(store.getState().projects[2].columnIds).toHaveLength(prev + 1)
  })
  test('MOVE_COLUMN', () => {
    store.dispatch({ type: '@@INIT' } as any)
    const PROJECT_ID = 0
    const COLUMN_ID = 1
    expect(store.getState().projects[PROJECT_ID].columnIds).toEqual([0, 1, 2])
    store.dispatch(moveColumnA(PROJECT_ID, COLUMN_ID, 'Back'))
    const columnIds = store.getState().projects[PROJECT_ID].columnIds
    expect(columnIds).toEqual([1, 0, 2])
    expect(store.getState()).toMatchSnapshot()
  })
})
