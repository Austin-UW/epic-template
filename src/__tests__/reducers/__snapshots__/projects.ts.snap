// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project reducer MOVE_COLUMN 1`] = `
Object {
  "columns": Object {
    "10": Object {
      "id": 10,
      "name": "In Progress",
      "taskIds": Array [],
    },
    "11": Object {
      "id": 11,
      "isCompletedColumn": true,
      "name": "Done",
      "taskIds": Array [],
    },
    "12": Object {
      "id": 12,
      "isCompletedColumn": true,
      "name": "NAME",
      "taskIds": Array [],
    },
    "3": Object {
      "id": 3,
      "name": "To-do",
      "taskIds": Array [
        6,
      ],
    },
    "4": Object {
      "id": 4,
      "name": "In Progress",
      "taskIds": Array [],
    },
    "5": Object {
      "id": 5,
      "isCompletedColumn": true,
      "name": "Complete",
      "taskIds": Array [],
    },
    "6": Object {
      "id": 6,
      "name": "To-do",
      "taskIds": Array [],
    },
    "7": Object {
      "id": 7,
      "name": "In Progress",
      "taskIds": Array [],
    },
    "8": Object {
      "id": 8,
      "isCompletedColumn": true,
      "name": "Complete",
      "taskIds": Array [],
    },
    "9": Object {
      "id": 9,
      "name": "To-do",
      "taskIds": Array [],
    },
  },
  "isLoading": false,
  "pastAction": Object {
    "type": "CHECK_TASK_OVERDUE",
  },
  "pomodoro": Object {
    "breakSeconds": 300,
    "currSeconds": 1500,
    "paused": true,
    "selectedTaskId": null,
    "selectingTask": false,
    "stopWatch": Object {
      "highest": 0,
      "pastTimes": Array [],
      "paused": true,
      "time": 0,
    },
    "totalTimes": Array [
      Object {
        "day": 0,
        "minutes": 0,
      },
      Object {
        "day": 1,
        "minutes": 0,
      },
      Object {
        "day": 2,
        "minutes": 0,
      },
      Object {
        "day": 3,
        "minutes": 0,
      },
      Object {
        "day": 4,
        "minutes": 0,
      },
      Object {
        "day": 5,
        "minutes": 0,
      },
      Object {
        "day": 6,
        "minutes": 0,
      },
    ],
    "workSeconds": 1500,
    "working": true,
  },
  "projects": Object {
    "0": Object {
      "categories": Array [
        "Doggo",
        "Weed",
        "Karen",
        "Great Ideas",
        "minimum wage",
        "cat",
      ],
      "columnIds": Array [
        1,
        0,
        2,
      ],
      "id": 0,
      "name": "Karen's Marijuana Dog Food",
    },
    "1": Object {
      "categories": Array [
        "Invasion",
        "Mass Genocide",
        "Destruction",
      ],
      "columnIds": Array [
        3,
        4,
        5,
      ],
      "id": 1,
      "name": "Russia is just a block of land and it should be ours",
    },
    "2": Object {
      "categories": Array [],
      "columnIds": Array [
        6,
        7,
        8,
      ],
      "id": 2,
      "name": "Believe in myself",
    },
  },
  "snackbar": Object {
    "message": "Katrina's Website created successfully!",
    "open": true,
    "variant": "success",
  },
  "tasks": Object {
    "6": Object {
      "color": "#FFFFFF",
      "id": 6,
      "name": "E",
      "subTasks": Array [],
      "timeWorkedOn": 0,
    },
  },
}
`;
